src/connection/activate.ts

```diff
@@ -0,0 +1,94 @@
+import { sendAnalyticsEvent } from '@uniswap/analytics'
+import { InterfaceEventName, WalletConnectionResult } from '@uniswap/analytics-events'
+import { sendEvent } from 'components/analytics'
+import { Connection } from 'connection'
+import { atom, useAtom } from 'jotai'
+import { useCallback } from 'react'
+import { useAppDispatch } from 'state/hooks'
+import { updateSelectedWallet } from 'state/user/reducer'
+
+import { didUserReject } from './utils'
+
+type PendingConnectionState = {
+  connection: Connection | undefined
+  error: any
+}
+
+const pendingConnectionAtom = atom<PendingConnectionState>({ connection: undefined, error: undefined })
+
+export function useActivateConnection() {
+  const dispatch = useAppDispatch()
+
+  const [pending, setPending] = useAtom(pendingConnectionAtom)
+
+  const updatePendingConnection = useCallback(
+    (connection: Connection | undefined) => {
+      // log selected wallet
+      if (connection !== undefined) {
+        sendEvent({
+          category: 'Wallet',
+          action: 'Change Wallet',
+          label: connection.type,
+        })
+      }
+      // Setting a new connection clears any pending errors
+      setPending({ connection, error: undefined })
+    },
+    [setPending]
+  )
+
+  const updatePendingError = useCallback(
+    (error: any) => {
+      setPending(({ connection }) => {
+        if (error !== undefined && connection !== undefined) {
+          sendAnalyticsEvent(InterfaceEventName.WALLET_CONNECT_TXN_COMPLETED, {
+            result: WalletConnectionResult.FAILED,
+            wallet_type: connection.getName(),
+          })
+        }
+
+        return { connection, error }
+      })
+    },
+    [setPending]
+  )
+
+  const cancelActivation = useCallback(async () => {
+    await pending.connection?.connector.deactivate?.()
+    updatePendingConnection(undefined)
+  }, [pending, updatePendingConnection])
+
+  const tryActivation = useCallback(
+    async (connection: Connection, onSuccess: () => void) => {
+      // Skips wallet connection if the connection should override the default behavior, i.e. install metamask or launch coinbase app
+      if (connection.overrideActivate?.()) return
```

Comment:
Could this be prefaced with a comment? What does this code do? If `overrideActivate` is present and returns a boolean, we return?

Reply:
I'm a bit confused, there is a prefacing comment right above it 😅 if overrideActivate exists and has an action described in the comment above to execute, then it returns true and we skip activation

Reply:
Yeah, I must have got confused by the comment! Overall, I am just curious when `connection.overrideActivate?.()` returns `false`, in that case, this will not early return!

Reply:
For most connection options, overrideActivate will be undefined. For something like MetaMask, we have to determine at runtime (b/c of injection timing) whether we should prompt metamask installation or not. So if metamaskConnection.overrideActivate?.() is called w/ injection present, false is returned.